src/H5VLcache_ext.c

```diff
@@ -5441,19 +5483,38 @@ static herr_t create_file_cache_on_local_storage(void *obj, void *file_args,
     file->H5DWMM->cache->duration = PERMANENT;
 
     if (file->H5LS->path != NULL) {
-      strcpy(file->H5DWMM->cache->path, file->H5LS->path);
-      strcat(file->H5DWMM->cache->path, "/");
-      strcat(file->H5DWMM->cache->path, basename((char *)name));
-      strcat(file->H5DWMM->cache->path, "-cache/");
-      // mkdir(file->H5DWMM->cache->path, 0755); // setup the folder with the
-      // name of the file, and put everything under it.
-
-      strcpy(file->H5DWMM->mmap->fname, file->H5DWMM->cache->path);
-      strcat(file->H5DWMM->mmap->fname, "mmap-");
+      // Build cache path: <storage_path>/<filename>-cache/
       char rnd[255];
       sprintf(rnd, "%d", file->H5DWMM->mpi->rank);
-      strcat(file->H5DWMM->mmap->fname, rnd);
-      strcat(file->H5DWMM->mmap->fname, ".dat");
+      char name_buf[4096];
+      strncpy(name_buf, name, sizeof(name_buf) - 1);
+      name_buf[sizeof(name_buf) - 1] = '\0';
+      char *base = basename(name_buf);
+
+      size_t path_len =
+          strlen(file->H5LS->path) + strlen(base) + strlen("-cache") + 2;
```

Comment:
Does this retain the trailing "/" at the end of the pathname? If logic was previously relying on that, it may be important to do so.